### PR TITLE
fix(eslint-plugin): [non-nullable-type-assertion-style] consider operator precedence when fixing

### LINKED ISSUE #7243: Bug: [non-nullable-type-assertion-style] autofix does not account for `await`
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.1.6&fileType=.tsx&code=CYUwxgNghgTiAEAzArgOzAFwJYHtVK1WABEoMoAKASgC54AFGHAWywGcQAeNjGQgc3gAfeKmQQIAPgDcAWABQCsHh7xmAD3gBeeFDYBPdPGp1GLdlx59U-SdrsBvBfHhwMyGPgpQA7lCwYBESk5NRUumzwVgJy8gC%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6JgeydqYXgEMARklpFStfsmSJo%2BSl1rIiSdFBnQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
declare function findData(): Promise<string | null>;

const mx = async () => {
  (await findData()) as string;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/non-nullable-type-assertion-style": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

```
declare function findData(): Promise<string | null>;

const mx = async (): Promise<string> => {
  return (await findData())!;
}
```

### Actual Result

```
declare function findData(): Promise<string | null>;

const mx = async (): Promise<string> => {
  return await findData()!;
}
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by auvred**
## PR Checklist

- [x] Addresses an existing open issue: fixes #7243
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

We need to wrap expression in parentheses if operator precedence is lower or equal to [`OperatorPrecedence.Unary`](https://github.com/typescript-eslint/typescript-eslint/blob/c79fec7e761cc2bccf839f761cff6f53652597f5/packages/eslint-plugin/src/util/getOperatorPrecedence.ts#L126-L139)

_Thanks to https://github.com/typescript-eslint/typescript-eslint/issues/7243#issuecomment-1637211868_

### CODE REVIEW

File: packages/eslint-plugin/tests/rules/non-nullable-type-assertion-style.test.ts [hunk 1]
```diff
@@ -199,6 +199,42 @@ declare const x: T;
 const y = x!;
       `,
     },
+    {
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/non-nullable-type-assertion-style.test.ts:202]: #6887 - could you split this in two please?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Changes look great, thanks! Just requesting a splitup of the tests. ✨

File: packages/eslint-plugin/tests/rules/non-nullable-type-assertion-style.test.ts (hunk 1 above)
Comment (auvred) [reply packages/eslint-plugin/tests/rules/non-nullable-type-assertion-style.test.ts:202]: Thank you for review, I've splitted it!

**Review by JoshuaKGoldberg — APPROVED** [review]
💯 thanks!

**Pull request merged**